src/core/wallet/Wallet.ts

```diff
@@ -214,104 +214,107 @@ class Wallet extends Channel<WalletEventName> {
+                      }
+                const balanceCall = isSettlementToken
+                    ? {
+                          contract: collateralTokenContract,
+                          contractName: ContractName.SETTLEMENT_TOKEN,
+                          funcName: "balanceOf",
+                          funcParams: [account],
+                      }
+                    : {
+                          contract: collateralTokenContract,
+                          contractName: ContractName.COLLATERAL_TOKENS,
+                          funcName: "balanceOf",
+                          funcParams: [account],
+                      }
+                const calls: ContractCall[] = [
+                    // NOTE: get decimals
+                    decimalsCall,
+                    // NOTE: get allowance
+                    allowanceCall,
+                    // NOTE: get balance
+                    balanceCall,
+                ]
+                const priceDecimalsCall = isSettlementToken
+                    ? undefined
+                    : {
+                          contract: collateralToken.priceFeedContract,
+                          contractName: ContractName.CHAINLINK_PRICE_FEED,
+                          funcName: "decimals",
+                          funcParams: [],
+                      }
+                const priceCall = isSettlementToken
+                    ? undefined
+                    : {
+                          contract: collateralToken.priceFeedContract,
+                          contractName: ContractName.CHAINLINK_PRICE_FEED,
+                          funcName: "getPrice",
+                          funcParams: [0],
+                      }
+                priceDecimalsCall && calls.push(priceDecimalsCall)
+                priceCall && calls.push(priceCall)
+                callsMap[`${collateralTokenAddress}`] = calls
+            })
+
+            const data = await multicall2.execute(Object.values(callsMap).flat())
+
+            const walletDataAll: WalletDataAll = {}
+            Object.entries(callsMap).forEach(([collateralAddress, calls]) => {
+                const dataChunk = data.splice(0, calls.length)
+                const decimals = dataChunk[0]
+                const allowance = bigNumber2BigAndScaleDown(dataChunk[1], decimals)
+                const balance = bigNumber2BigAndScaleDown(dataChunk[2], decimals)
+                const priceDecimals = dataChunk[3]
+                const price = dataChunk[4]
+                walletDataAll[`${collateralAddress}`] = {
+                    allowance,
+                    balance,
+                    // NOTE: SettlementToken price = 1
+                    price: priceDecimals && price ? bigNumber2BigAndScaleDown(price, priceDecimals).toNumber() : 1,
+                }
+            })
+
+            this.emit("updatedWalletDataAll", walletDataAll)
+        } catch (error) {
+            this.emit("updateError", { error })
```

Comment:
Should we add a class for this kind of error? like `MultiContractReadError`? instead of using `ContractReadeError` , which might make us easier to find errors on bugsnag ?

Reply:
Yes, I agree. Here we would probably encounter some errors during the parsing process as well. So let's add `MultiContractReadError` inside the MulticallReader. What do you think?